### ISSUE TITLE
Portability

### ISSUE THREAD

**Issue by fuchsia36** [wbenny/mini-tor#17]
This project could have a lot of use for various applications - the primary one I can think of would be to compile through emscripten for the browser, but also some uses embedded inside tools as a library. However, it's only available for win32.

How tightly coupled is it to the Windows platform? Would it be within its scope to try and make it portable? That is to say, putting the win32-specific code behind #ifdef directives and adding in alternatives for POSIX and whatnot.

From what I can see, the IO is abstracted in classes. Would porting be a 'simple' matter of editing these, or is there other coupling taking place?

As for the crypto, https://github.com/libtom/libtomcrypt might be relevant. I think it could be gutted to make a header-only library.

**Comment by wbenny** [comment]
Well, it is in the title - the OpenSSL is replaced by Microsoft's CNG/CryptoAPI. So it's a direct consequence of this project to be Win32-only. I've never intended to make `mini-tor` portable. The original Tor is portable enough. :)

**Comment by fuchsia36** [comment]
Right, that makes sense. But would it be feasible to port? For instance, test.cpp seems to touch win32 APIs directly, but I can't see anything in main.cpp. Is all of the windows-specific stuff hidden behind classes like how tcp_socket.cpp provides tcp_socket?

**Comment by wbenny** [comment]
`test.cpp` is just a scratch file that I use for various testing of the handcrafted STL-like algorithms etc. in the `mini` namespace. It probably even shouldn't be in the git, it might cause confusion.

Yes, the Win32-specific stuff is hardwired into the rest of the code. As I said, I didn't write it with portability in mind.

**Comment by cryptozoidberg** [comment]
Just in case anyone interested we made something like that but portable: https://github.com/hyle-team/tor-connect
